CIP-0000.md

```diff
@@ -0,0 +1,198 @@
+	* **Core** - improvements requiring a consensus fork, as well as changes that are not necessarily consensus critical but may be relevant to  “core dev” discussions, and the validator/node strategy.
+	* **Governance** - improvements to the governance proposal to the Celo platform including parameter changes. It should explain the reason for the change, enumerate the smart contract methods that will be called with parameters, and describe the impact and any risks. 
+	* **Networking** - improvements to networking protocol specifications such as devp2p, and proposed improvements like whisper and swarm.
+	* **Interface** - improvements around client [API/RPC] specifications and standards, and also certain language-level standards like method names and contract ABIs.
+	* **CRC** - (Celo [Request for Comments](https://en.wikipedia.org/wiki/Request_for_Comments)) application-level standards and conventions, including contract standards such as token standards, name registries, URI schemes, library/package formats, and wallet formats.
+* A **Meta CIP** describes a process surrounding Celo or proposes a change to (or an event in) a process. Process CIPs are like Standards Track CIPs but apply to areas other than the Celo protocol itself. They may propose an implementation, but not to Celo’s codebase; they often require community consensus; unlike Informational CIPs, they are more than recommendations, and users are typically not free to ignore them. Examples include code of conduct updates, procedures, guidelines, changes to the decision-making process, and changes to the tools or environment used in Celo development. Any Meta CIP is also considered a Process CIP.
+* An **Informational CIP** describes a Celo design issue, or provides general guidelines or information to the Celo community, but does not propose a new feature. Informational CIPs do not necessarily represent Celo community consensus or a recommendation, so users and implementers are free to ignore Informational CIPs or follow their advice.
+
+It is highly recommended that a single CIP contain a single key proposal or new idea.
+
+A CIP must meet certain minimum criteria. It must be a clear and complete description of the proposed enhancement. The enhancement must represent a net improvement as to be determined by the CIP editors during the PR review process. For more information on the CIP Editor role, see the appropriate section below.
+
+
+# CIP Workflow
+## Shepherding a CIP
+Parties involved in the process are you, the champion or _CIP author_, the CIP editors, and the Celo Core Developers (core developers are the developers working on the Celo client).
```

Comment:
In general, this document seems to be heavily focused on celo-client/celo-blockchain as the relevant domain, but I would suggest to expand the domain to the Celo Platform as a whole. Of course there are special considerations for celo-blockchain related CIPs (e.g. hard forks)

Reply:
can you be more specific about the scope expansion you're interested in? which repos and packages do you think should fall under the CIP process as opposed to being facilitated by a separate CGP process or not having a formal process?

Reply:
If we are looking at the whole Celo Platform like say Valora, it's probably not a good idea to involve it in the CIP process as it will greatly delay updates and fixes on Valora.

Reply:
I believe possible domains that could possibly fall under the CIP process that are not `celo-blockchain` are:
- The Celo Core Contracts (Governance, Proof-of-Stake, etc.) Examples include https://github.com/celo-org/celo-proposals/pull/17 or https://github.com/celo-org/celo-proposals/pull/42
- "core-protocol" primitives on the Celo platform (i.e. the identity protocol, including the on-chain `Attestations` contract but also off-chain components such as the `attestation-service` or `ODIS` and `ARKE`.)
- Another example would be offchain-storage, a la [CIP8](https://github.com/celo-org/celo-proposals/blob/master/CIPs/0008.md)

In general, we have a whole team at cLabs who is dedicated towards advancing all non-`celo-blockchain` related protocol efforts which we have tried to define as broadly speaking as building blocks for applications in the Celo ecosystem to build on. Of course the boundary towards applications can get blurry (would a wallet signer interface such as `WalletConnect` count?)

I'd love to hear how y'all think this view of the "Celo protocol" should (or shouldn't) be part of the CIP process

Reply:
+1 to all of these falling under the CIP process.

IIUC, the current proposal does allow for these to be covered

Core contracts could be *core* or *governance* (or both, if the implementation is *core* and the adoption is *governance*)
Similarly, other core primitives would be *core*
Standards for off-chain storage would be, IIUC, *crc*

Reply:
Yeah I agree that the current proposal can cover these, just some of the language in the proposal (like the above `core developers are the developers working on the Celo client`) might not feel the most inclusive in that regard

Reply:
I'd like to push back on ODIS and off-chain storage being part of the regular CIP process. These services are offered by specific entities, rather than core functionality of a distributed system. The goal of the core CIP process is to define the consensus protocol sufficiently that a second implementation could reach consensus with celo-blockchain. It is not to allow other people to use CLabs services like ODIS. Unless ODIS and CIP8 are critical to the continued function of the chain in a way I don't understand, they ought to be informational CIPs describing standards that use the Celo network (a la ERC20, EIP1967, etc), or not CIPs at all. I would like to exclude as many off-chain services as possible from this protocol. CIP8 certainly shouldn't be subject to the same review process as core CIPs that change consensus behavior. This process is slow, painful, and easily politicized.

While we're on the subject of consensus behavior, we should extend the CIP process to cover any smart contract that implements part of the consensus process or is called outside of the normal user-driven tx processing flow. The fee checks. The core contracts should be considered part of the `celo-blockchain` codebase, as they determine node behavior and can cause consensus failures between client versions

Reply:
I fully support @prestwich that off-chain services can be specified under non-Core category. Luckily, this CIP document has that specific category that applies to those off-chain services.

```
* **CRC** - (Celo [Request for Comments](https://en.wikipedia.org/wiki/Request_for_Comments)) application-level standards and conventions, including contract standards such as token standards, name registries, URI schemes, library/package formats, and wallet formats.
```

We can use the CRC category proposed for any service level standards off-chain.

The main goal of "Core" is to be able to specify the Celo Standard, just as "Core" for EIPs specifies the Ethereum Standard, and the main benefits are outlined by James. Long term, I hope it allows the generation of a yellow paper defining the precise specification of the Celo Standard. 

I also agree that for smart contracts, if they're defining behavior that affects how consensus is reached and other behavior like activation of a network upgrade from a specific hard fork block or the activation of a specific gas limit, then they are under the Core Category of the Standards track.

With CRCs, we can include as many off-chain services that need standardization as possible. I can help everyone draft them if needed, and we can discuss them in our debut calls for the CIP process.

Reply:
> Unless ODIS and CIP8 are critical to the continued function of the chain in a way I don't understand, they ought to be informational CIPs describing standards that use the Celo network (a la ERC20, EIP1967, etc)

I agree with this wrt CIP8. RE ODIS, I guess it's a question of this off-chain service is considered part of e.g. the attestation protocol primitive.

> While we're on the subject of consensus behavior, we should extend the CIP process to cover any smart contract that implements part of the consensus process or is called outside of the normal user-driven tx processing flow

> I also agree that for smart contracts, if they're defining behavior that affects how consensus is reached and other behavior like activation of a network upgrade from a specific hard fork block or the activation of a specific gas limit, then they are under the Core Category of the Standards track.

I agree with these statements and would suggest extending it to cover the stability and attestation protocols, despite not necessarily affecting consensus or node behavior.

One open question I have is how we treat upgrades to core contracts under the proposed process. Seemingly, proposed changes should go through the CIP process before being merged to master, but also these changes will need to be proposed and adopted via on-chain governance.

Reply:
My view is that one of our primary goals should be to ensure that apps like Valora built on top of the Celo platform can inter-operate, and that when a new feature/service/process is required/desired, it is low-friction to have that adopted as part of the Celo Platform.  In fact, the layer where we have seen most need for a standards process is around things like identity, comment encryption, QR code formats, etc - because the loss of value to the platform in having approaches to these things diverge is huge. Clearly services like ODIS test that "Celo Platform" boundary but these should not be seen as cLabs services in the long term, just like the oracles.  At the very minimum, any contract owned by Governance should be part of the CIP process.

For me, it is a distant secondary goal to have a second clean-room implementation of celo blockchain. Many successful open source projects are "standards by reference implementation" and further implementations reduce development velocity.

Reply:
I redefined it as `core developers are the developers working on the Celo Platform`.